src/main/scala/loamstream/apps/LoamRunner.scala

```diff
@@ -36,54 +30,13 @@ object LoamRunner {
       info(compilationResults.summary)
 
       if (compilationResults.isValid && compilationResults.isSuccess) {
-        Right(process(compilationResults.graphSource))
+        compilationResults.contextOpt.map(_.graph) match {
+          case Some(graph) => Right(loamEngine.run(graph))
+          case None => Left(compilationResults)
+        }
```

Comment:
Sorry, not quite following this code. It's straight forward, but what I'm not following is that it returns an `Either`, but the `Left` side is used if the complication is invalid, not successful, or .... has no graph? Just want to make sure I'm understanding the failure cases as it looks like it can return a failure from a successful `if` branch.  I'm wondering if perhaps it might be beneficial to have an enumeration of results?

Reply:
You got it right.  "Compilation" is a bit of a loaded term here, since it means actually compiling (with `scalac`) .loam files (or rather, .scala files derived from them) _and then_ evaluating the compiled code.  (Using 'compilation' in this way was not my idea.)  

The idea here is that `compilationResults` contains information from `scalac` as well as any `LoamGraph` that resulted from evaluating the .loam code.  Evaluating .loam code that doesn't do anything is deemed an error, though one could argue that that should be allowed.

I'm hesitant to add one more enum, but a general overhaul of the ".loam files => `LoamGraph`" process, particularly regarding how results and errors are reported to users is a worthwhile thing.

Reply:
Yeah, when to add another enumeration can be tricky. I'd ignore it for now then.

Reply:
You won me over with the ADT idea.  It makes things simpler.